### PR TITLE
feat: smt checker updates

### LINKED ISSUE #578: SMTChecker
We should do the initial setup of this https://docs.soliditylang.org/en/v0.8.19/smtchecker.html

I presume there will be modifications we need to do to the code to get this to work, those can be in separate bounties after this is feature is set up.

https://twitter.com/paulrberg/status/1627654614811115520?s=46&t=lQUXMD9A0xd14FiMGl8IEQ

### PR DESCRIPTION
**Pull request by FibrinLab**
Resolves #578

### CODE REVIEW

**Comment by FibrinLab** [comment]
What has been done so far.

1. Create a separate script like "yarn run build:smt" which only runs smt check. 
2. Foundry.toml modified via a js script to add profile.default.model_checker option.
3. Github workflow created that runs smt check only onworkflow_dispatch (not on every PR, because build time is increased greatly).

Thanks
@pavlovcik @rndquu

File: .github/workflows/smt-checker.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+name: SMT Checker
+
+on: workflow_dispatch
+
+jobs:
+  build-and-test:
+    name: Build & Test
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Pls use more appropriate name

File: .github/workflows/smt-checker.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: read
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: You sure we need this permission for smt workflow?

File: .github/workflows/smt-checker.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Setup Node
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Unnecessary step

File: .github/workflows/smt-checker.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        uses: actions/setup-node@v3
+        with:
+          node-version: "18.14.1"
+
+      - name: Install dependencies
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Unnecessary step

File: .github/workflows/smt-checker.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        run: sudo apt -y update && sudo apt -y install libusb-1.0-0-dev libudev-dev
+
+      - name: Setup Foundry
+        uses: foundry-rs/foundry-toolchain@v1
+        with:
+          version: nightly-87bc53fc6c874bd4c92d97ed180b949e3a36d78c
+
+      - name: Yarn Install
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml:21]: Pls leave only the steps required to run SMT checker

File: .github/workflows/smt-checker.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        run: yarn install --mode=skip-build && yarn allow-scripts
+
+      - name: Set Environment Variables
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Unnecessary step

File: .github/workflows/smt-checker.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+          echo "INFURA_ID=$INFURA_ID" > ./packages/dapp/.env
+          echo "API_KEY_ALCHEMY=$API_KEY_ALCHEMY" >> ./packages/dapp/.env
+
+      - name: Setup Forge
+        run: yarn workspace @ubiquity/contracts forge:install
+
+      - name: Build and Run SMT Check
+        run: yarn build:smt
+
+      - name: Test All
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Unnecessary step

File: .github/workflows/smt-checker.yml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        run: yarn test:all
+
+      - name: Prepare Deployment Artifact
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Unnecessary step

File: .github/workflows/smt-checker.yml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+          echo "event_name=$EVENT_NAME,pr_number=$PR_NUMBER,sha=$SHA"
+          mkdir -p ./pr
+          echo "event_name=$EVENT_NAME,pr_number=$PR_NUMBER,sha=$SHA," > ./pr/pr_number
+          cd ./packages/dapp && zip -r pull-request.zip ./dist && cp pull-request.zip ../../pr
+      - uses: actions/upload-artifact@v3
+        with:
+          name: pr
+          path: pr/
+
+      - name: Run Server
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml]: Unnecessary step

File: .github/workflows/smt-checker.yml [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+      - name: Set Environment Variables
+        env:
+          INFURA_ID: ${{ secrets.INFURA_ID }}
+          API_KEY_ALCHEMY: ${{ secrets.API_KEY_ALCHEMY }}
+        run: |
+          echo "INFURA_ID=$INFURA_ID" > ./packages/dapp/.env
+          echo "API_KEY_ALCHEMY=$API_KEY_ALCHEMY" >> ./packages/dapp/.env
+
+      - name: Setup Forge
```
Comment (rndquu) [inline .github/workflows/smt-checker.yml:24]: Probably an unnecessary step

File: packages/contracts/scripts/prebuild/smtchecker_prebuild.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    contract.replace("../../", "")
+  );
+
+  let contractsString = "";
+  for (const contract of contracts) {
+    const contractContent = fs.readFileSync(
+      path.join(__dirname, "../../", contract),
+      "utf8"
+    );
+    const namesMatched = contractContent.match(/contract\s+(\w+)\s+(is|{)/g);
```
Comment (rndquu) [inline packages/contracts/scripts/prebuild/smtchecker_prebuild.ts:37]: Perhaps it is better to use glob as @zgorizzo69 stated here https://github.com/ubiquity/ubiquity-dollar/pull/604#discussion_r1138332187 ? Searching in file content is brittle

**Review by rndquu — CHANGES_REQUESTED** [review]
@FibrinLab 

Thank you for the PR.

Questions:
1. There are no `targets` set in this PR like here https://github.com/ubiquity/ubiquity-dollar/pull/604/files#diff-5084a3d2809ef6c86199f46d1d985964c92c3a98247da067d6cb04b18b0a40a9R30 . How is it working without the targets?
2. Did you run the script? How long did it take to finish a full SMT check?

File: .github/workflows/smt-checker.yml [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    runs-on: ubuntu-22.04
+    
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Setup Node
+        uses: actions/setup-node@v3
+        with:
+          node-version: "18.14.1"
```
Comment (pavlovcik) [inline .github/workflows/smt-checker.yml:19]:
```suggestion
```

File: .github/workflows/smt-checker.yml
Comment (0x4007) [inline .github/workflows/smt-checker.yml:1]: It doesn't look like you incorporated @rndquu's input. You should work on removing the unnecessary steps.

File: packages/contracts/package.json [hunk 13]
```diff
@@ -34,6 +34,7 @@
     "clean:yarn": "yarn clean",
     "clean:forge": "forge clean",
     "build": "forge build",
+    "build:smt": "tsx ./scripts/prebuild/smtchecker_prebuild.ts && forge build",
```
Comment (0x4007) [inline packages/contracts/package.json:37]: You should use use `npm-run-all` in order to stay consistent with how we do it in this same package.json for the other scripts. e.g. `run-s build:smt build`

Also use kebab case for the file names. So `smtchecker-prebuild.ts` for example.

File: packages/contracts/scripts/prebuild/smtchecker_prebuild.ts [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    } else {
+      arrayOfFiles.push(filePath);
+    }
+  }
+
+  return arrayOfFiles;
+}
+
+function updateFoundryFile() {
+  const fileContent = fs.readFileSync(foundry_file, "utf8");
```
Comment (0x4007) [inline packages/contracts/scripts/prebuild/smtchecker_prebuild.ts]:
```suggestion
  const fileContent = fs.readFileSync(foundryFile, "utf8");
```

**Review by 0x4007 — CHANGES_REQUESTED** [review]
It doesn't look like you incorporated @rndquu's input. You should work on removing the unnecessary steps in the CI. Also I included some code style feedback. Camel casing variables and kebab casing filenames!

File: packages/contracts/foundry.toml [hunk 15]
```diff
@@ -26,12 +26,14 @@ mainnet = "https://eth.ubq.fi/v1/mainnet"
 
 [profile.default.model_checker]
 engine = 'chc'
-timeout = 10_000
```
Comment (0x4007) [inline packages/contracts/foundry.toml]: Why the change?

File: packages/contracts/scripts/prebuild/smtchecker_prebuild.ts [hunk 16]
```diff
@@ -3,50 +3,52 @@ import path from "path";
...
-function getAllFiles(dirPath: string, arrayOfFiles: string[]): string[] {
-  const fullPath = path.resolve(dirPath);
+export const getAllFiles = (dirPath: string): string[] => {
+  const fullPath = path.join(__dirname, dirPath);
+
   const files = fs.readdirSync(fullPath);
+  let arrayOfFiles: string[] = [];
 
-  for (const file of files) {
+  files.forEach((file) => {
```
Comment (0x4007) [inline packages/contracts/scripts/prebuild/smtchecker_prebuild.ts:12]: Also curious about this change why its necessary.

File: packages/contracts/scripts/prebuild/smtchecker_prebuild.ts [hunk 17]
```diff
@@ -3,50 +3,52 @@ import path from "path";
...
-      const names = namesMatched.map((name) =>
-        name.replace(/contract|{|is/g, "").trim()
-      );
+      const names = namesMatched.map((name: string) => name.replace("contract", "").replace(/{/g, "").replace("is", "").trim());
       contractsString += `'${contract}' = [ '${names.join(", ")}' ]\n`;
     }
-  }
+  });
 
+  /// Problem is from here
```
Comment (0x4007) [inline packages/contracts/scripts/prebuild/smtchecker_prebuild.ts:44]: What is the problem?

**Review by 0x4007 — COMMENTED** [review]
It seems like there were a lot of unnecessary refactors. Aside from that, it doesn't appear that you resolved @rndquu's comment.

**Comment by FibrinLab** [comment]
> It seems like there were a lot of unnecessary refactors. Aside from that, it doesn't appear that you resolved @rndquu's comment.

I had to keep some steps so the CI runs. I am making a final commit

**Comment by sergfeldman** [comment]
As I see, @FibrinLab had not completed this bounty on time and was unassigned https://github.com/ubiquity/ubiquity-dollar/issues/578

@pavlovcik @rndquu Should we close this PR then?

**Comment by 0x4007** [comment]
I think that makes sense to close the pull requests when unassigned. Perhaps we should also have the bot do that.

**Pull request closed without merging**